src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -0,0 +1,173 @@
+package fs2.async.mutable
+
+import fs2._
+import fs2.async.AsyncExt.Change
+
+import fs2.async.{immutable, AsyncExt}
+
+
+
+/**
+ * Asynchronous queue interface. Operations are all nonblocking in their
+ * implementations, but may be 'semantically' blocking. For instance,
+ * a queue may have a bound on its size, in which case enqueuing may
+ * block until there is an offsetting dequeue.
+ */
+trait Queue[F[_],A] {
+
+
+  /**
+   * Enqueue one element in this `Queue`.
+   *
+   * Evaluates to `false` if the queue is full, indicating the `a` was not queued up.
+   * Evaluates to `true` if the `a` was queued up successfully.
+   *
+   * @param a `A` to enqueue
+   */
+  def enqueueOne(a: A): F[Boolean]
+
+  /**
+   * Provides a process that dequeue from this queue.
+   * When multiple consumers dequeue from this queue,
+   * they dequeue in first-come, first-serve order.
+   *
+   * Please use `Topic` instead of `Queue` when all subscribers
+   * need to see each value enqueued.
+   *
+   * This process is equivalent to `dequeueBatch(1)`.
+   */
+  def dequeue: Stream[F, A]
+
+  /**
+   * The time-varying size of this `Queue`. This signal refreshes
+   * only when size changes. Offsetting enqueues and de-queues may
+   * not result in refreshes.
+   */
+  def size: Stream[F,immutable.Signal[F,Int]]
+
+  /**
+   * The size bound on the queue.
+   * Returns None if the queue is unbounded.
+   */
+  def upperBound: Option[Int]
+
+  /**
+   * Returns the available number of entries in the queue.
+   * Always returns `Int.MaxValue` when the queue is unbounded.
+   */
+  def available: Stream[F,immutable.Signal[F,Int]]
+
+  /**
+   * Returns `true` when the queue has reached its upper size bound.
+   * Always returns `false` when the queue is unbounded.
+   */
+  def full: Stream[F,immutable.Signal[F,Boolean]]
+
+
+
+
+}
+
+object Queue {
+
+    private type DeQueue[F[_],A] = (A => F[Unit])
+
+
+  /**
+    * Internal state of the queue
+    * @param queue    Queue, expressed as vector for fast cons/uncons from head/tail
+    * @param deq      A list of De-Queuer (filled up when queue is empty)
+    * @tparam F
+    * @tparam A
+    */
+    private case class State[F[_],A](
```

Comment:
To allow for the blocking `enqueue1` operation I mentioned above, I'd suggest maybe:

```Scala
// Type enforces invariants that: 
// a) Can't be any dequeues waiting unless queue is empty (if queue were nonempty we'd have dispatched those waiting)
// b) Can't be any enqueues waiting unless queue is full
// when adding to a Nonempty, check if this transitions to Full state
trait State[F[_],A]
case class Empty[F[_],A](dequeues: Vector[DeQueue[F,A]]) extends State[F,A]
case class Nonempty[F[_],A](q: Vector[A]) extends State[F,A]
case class Full[F[_],A](q: Vector[A], enqueues: ???) extends State[F,A] // not sure type of `enqueues`
```